### ISSUE TITLE
Client-side file storage

### ISSUE THREAD

**Issue by WillConker** [minetest/minetest#14821]
### Problem

Currently client-side mods cannot read or write files on the client.
This is inconvenient for me as I am trying to write a CSM which sends skins to the server, but the only way (as far as I can see) to read the skin data in the CSM is to call `skin_data = dofile(skin.lua)`, where `skin.lua` contains `return <base64 encoded skin>`

This is obviously very hacky.
I could see this feature also being useful for mods such as the chatlog csm (https://forum.minetest.net/viewtopic.php?t=17052)

### Solutions

Now that the `mod_data` directory is supported in 5.9.0, also add a `clientmod_data` directory where client mods can read  and write files. Of course, the reading and writing of files by client-side mods should still be controlled by mod security, just like server-side mods.

### Alternatives

I can't think of any good alternatives to this - an (inferior) alternative solution would be to allow CSMs to read their mod directory, but not doing this is (afaics) the whole reason `mod_data` was introduced in the first place (as updates overwrite stored data).

**Comment by WillConker** [comment]
Couldn't the provided `io` global just be set to a different lua module which fakes file writing in the browser (if browser used) while having the same api? That would make the api simpler to understand as it would be the same as server-side

> This needs a disk space limit.

This makes sense, why isn't this enforced for server-side mods?

**Comment by sfan5** [comment]
> This makes sense, why isn't this enforced for server-side mods?

The long term plan is for the client to run code sent by the server (like JS in the browser). This needs much more careful sandboxing and handling than server mods, which are always installed explicitly by wish of the user.

**Comment by WillConker** [comment]
> > This makes sense, why isn't this enforced for server-side mods?
> 
> The long term plan is for the client to run code sent by the server (like JS in the browser). This needs much more careful sandboxing and handling than server mods, which are always installed explicitly by wish of the user.

This would be really nice, but given that we're not getting SSCSM any time soon (afaics), this feature shouldn't have to support it.

**Comment by sfan5** [comment]
I mean sure, we can add features that are incompatible with SSCSM but who's going to take the user complaints when we inevitably have to remove it again?

**Comment by WillConker** [comment]
> I mean sure, we can add features that are incompatible with SSCSM but who's going to take the user complaints when we inevitably have to remove it again?

No, I mean that non-SSCSMs should get a different api to SSCSMs, which includes read/write to filesystem. This is easy to restrict later (can be added with all the other SSCSM restrictions) and requires no extra effort before SSCSMs come along.

**Comment by WillConker** [comment]
> There's also another reason to abstract storage and not use file IO (or at least emulate the IO): security sandboxes like pledge(2) can forbid opening new file descriptors while permitting operations on already opened ones, such as a database file.

I don't understand, are CSMs run in some sandbox different to server mods?

**Comment by WillConker** [comment]
> @WillConker No, a server admin might want to place MT in a sandbox like this for extra security. Ideally, if MT itself had a policy of not allowing arbitrary IO access (data directories are a step towards that), it could use those extra sandbox features on systems that support them, instead of having users do it manually with FFI. The less arbitrary IO we allow, the stricter such a sandbox can be without breaking anything.

This isn't implemented yet for server-side mods - they write directly to filesystem (albeit controlled by mod security). So why can't we have data directories for client-side mods for now, then when 'abstraction' gets implemented like you suggest, it could be implemented for both client- and server- mods?

**Comment by WillConker** [comment]
Ok, well I definitely can't argue with the fact that the lua `io` library is annoying.

> ```
> local storage = minetest.get_mod_storage();
> storage.x = storage.x or {}; -- create directory
> storage.x.duck = "quack"; -- create or replace file
> ```

I guess an abstracted system is ok, if we change it to do what this code is meant to do. At the moment that code creates an entry in a json file, not an actual directory and file inside. This prohibits easy modification of mod data by users (i.e. in my case adding their own skin file to mod data)

**Comment by WillConker** [comment]
Given that, I suppose a better feature request would be something like:
- Add a `clientmod_data` directory for CSMs to have world-independent mod storage
- CSMs would not be allowed to access it directly but would instead have to use an abstracted system like the one @hecktest describes above.
- This would be a userdata object (with a name such as `ModDataRef`) but would behave like a lua table, apart from that:
  - All keys must be generally valid filenames (i.e. no weird characters)
  - When assigning to a key, the value assigned must be either a lua table (representing a directory) which would be converted to a child `ModDataRef` when assigned, or a lua string (representing the bytes of a file)
  - When anything is assigned it is immediately written to the filesystem (or stored otherwise) by default.
- Where possible (i.e. whenever there is an actual filesystem), the structure of the `clientmod_data` would be the same as assigned in the lua code, so that users could more easily interact with their mods
- The `ModDataRef` (or whatever you want to call it) object would support a way to get all of its child filestrings and `ModDataRef`s (directories)

**Comment by hlqkj** [comment]
> > > This makes sense, why isn't this enforced for server-side mods?
> > 
> > 
> > The long term plan is for the client to run code sent by the server (like JS in the browser). This needs much more careful sandboxing and handling than server mods, which are always installed explicitly by wish of the user.
> 
> This would be really nice, but given that we're not getting SSCSM any time soon (afaics), this feature shouldn't have to support it.
> 
> **edit: by which I mean, currently csms can be trusted to the same extent as server mods because they must be installed manually. I think if sscsm ever does get introduced, it should be more restricted than client-installed csms**

Sorry if I "relive" this after three weeks. I just want to sharpen and support the above "edit": there are actually **no reasons** why a CSM (under the present situation) should be less trusted than a regular server mod in any way: it's _me_ who _willingly_ install or even develop a CSM, and thus I should legit have a full set of API to script with, rather than being sandboxed by my own machine.

The "debate" in this case, should only be about which features to limit as to prevent cheating, RE. CSM restriction flags.

As an example, the current silly file access sandbox made it impossible for me to develop a decent client-side chat logging mod, which is much needed. (See also: https://github.com/minetest/minetest/pull/14289.) Or to have a CSM live-reloaded like I regularly do with server mods, which would make CSM dev so much faster. But these are just my own use-cases, the skin upload above proves there are plenty

User-provided and Server-Sent CSM are different things, and as such different levels of sandboxing should be considered.

A user provided CSM should have unrestricted local access, while a SS-CSM should be restricted as suggested by above comments.
